src/main/java/org/truffleruby/language/arguments/CheckKeywordArgumentNode.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.arguments;
+
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import org.truffleruby.core.hash.RubyHash;
+import org.truffleruby.core.hash.library.HashStoreLibrary;
+import org.truffleruby.core.symbol.RubySymbol;
+import org.truffleruby.language.RubyContextSourceNode;
+import org.truffleruby.language.RubyNode;
+import org.truffleruby.language.arguments.keywords.KeywordDescriptor;
+import org.truffleruby.language.arguments.keywords.NonEmptyKeywordDescriptor;
+import org.truffleruby.language.locals.WriteLocalVariableNode;
+
+public class CheckKeywordArgumentNode extends RubyContextSourceNode {
+
+    public static final CheckKeywordArgumentNode[] EMPTY_ARRAY = new CheckKeywordArgumentNode[0];
+
+    private final FrameSlot slot;
+    private final RubySymbol name;
+
+    @Child protected ReadUserKeywordsHashNode readUserKeywordsHashNode;
+    @Child protected HashStoreLibrary hashStoreLibrary = HashStoreLibrary.createDispatched();
+    @Child protected WriteLocalVariableNode writeDefaultNode;
+
+    public CheckKeywordArgumentNode(FrameSlot slot, int minimum, RubyNode defaultValue, RubySymbol name) {
+        this.slot = slot;
+        this.name = name;
+        writeDefaultNode = new WriteLocalVariableNode(slot, defaultValue);
+        readUserKeywordsHashNode = new ReadUserKeywordsHashNode(minimum);
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        return execute(frame, RubyArguments.getKeywordArgumentsDescriptorUnsafe(frame));
+    }
+
+    public Object execute(VirtualFrame frame, KeywordDescriptor descriptor) {
+        // Get the current value of the parameter.
+
+        final Object value = frame.getValue(slot);
+
+        // If it has a value, then we're done.
+
+        if (value != ReadArgumentsNode.MISSING) {
```

Comment:
This check is not optimal and will only go away if we profiled the type of `value`.
It would be better to use the descriptor here, that tells us exactly which keywords were provided and which were not.

Reply:
Right.

The original thinking was to have the code for assigning missing optionals done once, for all specialisations, to stop default value expressions being duplicated. But I looked around and 99% of optionals are literals.

@nirvdrum

Reply:
@eregon Using the descriptor directly is straightforward. Are you suggesting the addition of a condition or branch profile as well?

Reply:
We shouldn't need profile, since both the descriptor and the set of keywords are PE constants and so the compiler can compute all that.

Reply:
Since the KeywordDescriptor doesn't contain all keys we can't really do matching on PE constants only.
We could probably avoid the default value here and directly handle keyword present ? use value : execute default.